src/realm/object-store/sync/async_open_task.cpp

```diff
@@ -66,7 +67,35 @@ void AsyncOpenTask::start(util::UniqueFunction<void(ThreadSafeReference, std::ex
         catch (...) {
             return callback({}, std::current_exception());
         }
-        callback(std::move(realm), nullptr);
+
+        // TODO: add a config param to guard this
+        // before the to
+        // the file has been downloaded.
+        SharedRealm shared_realm = Realm::get_shared_realm(std::move(realm));
+        auto config = shared_realm->config();
+
+        if (config.sync_config && config.sync_config->flx_sync_requested) {
+            using namespace sync;
+            // I am assuming that the latest subscription at this point is running the initial query
```

Comment:
A fresh realm will have no subscription sets other than the default one (zero'th). This is marked Complete when schema is downloaded from the server (done by the time wait_for_download_completion's callback is called), therefore `get_state_change_notification` below returns immediately.

We need a way to pass the subscriptions, but creating subscriptions require access to a Realm instance. So my take is to have a new callback which takes a Realm as parameter and returns a SubscriptionSet/MutableSubscriptionSet/Query[] which we can use to create the initial subscription set. We then commit it and wait for Complete state.